K8SPSMDB-200: Add support for service annotations

jasper-d opened the pull request:
[![K8SPSMDB-200](https://badgen.net/badge/JIRA/K8SPSMDB-200/green)](https://jira.percona.com/browse/K8SPSMDB-200)<!-- PR-BADGE: PLEASE DO NOT REMOVE THIS COMMENT -->

Adds support for service annotations and external-dns annotation.

- [x] Test with external-dns
- [ ] ~e2e tests are failing on my minikube (even w/o my changes), not sure how to execute them and what environment is needed~

See https://github.com/percona/percona-server-mongodb-operator/pull/361#issuecomment-634222147 for update notes.

it-percona commented:
[![CLA assistant check](https://cla-assistant.percona.com:443/pull/badge/not_signed)](https://cla-assistant.percona.com:443/percona/percona-server-mongodb-operator?pullRequest=361) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.percona.com:443/percona/percona-server-mongodb-operator?pullRequest=361) before we can accept your contribution.<br/><hr/>**Jasper Dunker** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.percona.com:443/check/percona/percona-server-mongodb-operator?pullRequest=361) it.</sub>

JNKPercona commented:
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| init-deploy | passed |
| one-pod | passed |
| upgrade | passed |
| self-healing | failed |
| monitoring | passed |
| limits | passed |
| operator-self-healing | failed |
| upgrade-consistency | failed |
| scaling | passed |
| demand-backup | failed |
| scheduled-backup | failed |
| monitoring-2-0 | failed |
| security-context | passed |
| arbiter | passed |
| service-per-pod | failed |
| liveness | passed |

JNKPercona commented:
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| self-healing | passed |
| init-deploy | passed |
| limits | passed |
| scaling | passed |
| security-context | passed |
| upgrade | passed |
| upgrade-consistency | passed |
| one-pod | passed |
| monitoring | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| arbiter | passed |
| service-per-pod | failed |
| operator-self-healing | passed |
| scheduled-backup | passed |
| liveness | passed |

jasper-d commented:
@delgod I encountered a few issues when testing on Azure AKS. First I hadn't though about FW rules. I added LoadBalancerAllowedSourceRanges in 1f4e610 to allow traffic from selected hosts when using LoadBalancer as the service type (at least in our Azure setup the default deny all policy will otherwise block access to mongo).
I also had problems getting a MongoDB cluster up and running until I realized that when `expose.enabled: true` is used, no headless service is created. This apparently required additional source  ranges to enable access for the operator through the LoadBalancers.

Personally, I don't see a benefit in removing the headless service which also provides SRV lookups for cluster-internal clients, simplifying MongoDB connection strings. I added a third commit (1433851) to always create a headlless service, even when `expose.enabled: true` and to use that service for communication from the operator to the cluster. This is probably out of scope of this PR and there are probably good reasons to scrap the headless service, so I don't mind to revert it if you wish so.

However, I think it would be good to at least have a choice to run a headless service and external services side by side (maybe through an option like `expose.keepHeadlessService: bool`). But I think that could be done in a different PR.

JNKPercona commented:
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | failed |
| storage | passed |
| one-pod | failed |
| limits | failed |
| upgrade | failed |
| scaling | passed |
| self-healing | passed |
| monitoring | passed |
| upgrade-consistency | passed |
| security-context | passed |
| demand-backup | failed |
| operator-self-healing | passed |
| monitoring-2-0 | passed |
| scheduled-backup | passed |
| arbiter | passed |
| service-per-pod | failed |
| liveness | passed |

JNKPercona commented:
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | passed |
| one-pod | passed |
| upgrade | failed |
| self-healing | passed |
| monitoring | passed |
| limits | passed |
| scaling | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| security-context | passed |
| operator-self-healing | passed |
| arbiter | passed |
| scheduled-backup | failed |
| service-per-pod | failed |
| liveness | passed |

delgod commented on pkg/psmdb/service.go:
```diff
@@ -91,6 +101,11 @@ func ExternalService(m *api.PerconaServerMongoDB, replset *api.ReplsetSpec, podN
 	case corev1.ServiceTypeLoadBalancer:
 		svc.Spec.Type = corev1.ServiceTypeLoadBalancer
 		svc.Spec.ExternalTrafficPolicy = "Cluster"
+		if len(replset.Expose.LoadBalancerSourceRanges) > 0 {
+			for _, sr := range replset.Expose.LoadBalancerSourceRanges {
+				svc.Spec.LoadBalancerSourceRanges = append(svc.Spec.LoadBalancerSourceRanges, sr)
+			}
+		}
```
can we just assign like this `svc.Spec.LoadBalancerSourceRanges = replset.Expose.LoadBalancerSourceRanges` ?

jasper-d replied:
Yes, done

delgod commented on pkg/controller/perconaservermongodb/psmdb_controller.go:
```diff
@@ -250,7 +250,25 @@ func (r *ReconcilePerconaServerMongoDB) Reconcile(request reconcile.Request) (re
 			return reconcile.Result{}, err
 		}
 
-		// Create Service
```
just Q: why it was needed to move?

jasper-d replied:
Not needed, but I felt it would make more sense to move creation of the headless service up here because it is now always done while the external service creation is still conditional. Don't mind to move it down if you prefer it.

delgod commented on pkg/psmdb/service.go:
```diff
@@ -185,45 +200,20 @@ func getIngressPoint(pod corev1.Pod, cl client.Client) (string, error) {
 
 // MongoHost returns the mongo host for given pod
-func MongoHost(cl client.Client, m *api.PerconaServerMongoDB, replset *api.ReplsetSpec, pod corev1.Pod) (string, error) {
-	if replset.Expose.Enabled {
```
can we analyse ExternalDnsZone
1) if ExternalDnsZone set - return `fmt.Sprintf("%s.%s", podName, replset.Expose.ExternalDnsZone)`
2) if ExternalDnsZone = "" then `return getExtAddr(cl, m.Namespace, pod)` 
?

jasper-d replied:
That would mean that MongoHost would always return the service when enabled, either using the hostname or IP of the service, right?
Since MongoHost is used to get get the host name for cluster internal communication (i.e. operator to MongoDb) I think using the services complicates things here (communication would go from inside the cluster through the service just to be routed back to the Mongo pods). This would also require additional FW rules to allow the operator pod to connect to the Mongo pods through that service when using LoadBalancer.
Maybe I'm missing something?

delgod replied:
> That would mean that MongoHost would always return the service when enabled, either using the hostname or IP of the service, right?

Idea is to analyze ExternalDnsZone, if it is empty, that code will behave as before adding ExternalDnsZone feature. if it is set then use ExternalDnsZone way.

>  I think using the services complicates things here

yes, you right, but it is current functionality and we don't want to delete current feature.

jasper-d replied:
Alright, changed it back to the previous behaviour and added the requested change w.r.t. ExternalDnsZone.
I understand that you don't want to change the behaviour right now for compatibility reasons, but would it be an option to either change it at a later point or add an additional configuration that allows users to opt out of using external service for this?

jasper-d commented:
I‘m on vocation and will take a look next week.

delgod commented:
@jasper-d could you please rebase on the current master?

JNKPercona commented:
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| storage | passed |
| init-deploy | passed |
| upgrade | passed |
| self-healing | failed |
| monitoring | passed |
| limits | passed |
| operator-self-healing | failed |
| upgrade-consistency | passed |
| scaling | passed |
| monitoring-2-0 | failed |
| security-context | passed |
| arbiter | passed |
| demand-backup | passed |
| service-per-pod | failed |
| scheduled-backup | passed |
| liveness | passed |

JNKPercona commented:
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | failed |
| storage | passed |
| upgrade | passed |
| one-pod | failed |
| limits | passed |
| self-healing | passed |
| scaling | failed |
| monitoring | failed |
| upgrade-consistency | passed |
| security-context | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| operator-self-healing | passed |
| scheduled-backup | passed |
| arbiter | passed |
| service-per-pod | failed |
| liveness | failed |

JNKPercona commented:
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| upgrade | failed |
| init-deploy | passed |
| one-pod | passed |
| self-healing | failed |
| monitoring | passed |
| upgrade-consistency | failed |
| limits | failed |
| operator-self-healing | passed |
| demand-backup | passed |
| scaling | failed |
| monitoring-2-0 | passed |
| security-context | passed |
| scheduled-backup | passed |
| arbiter | passed |
| service-per-pod | failed |
| liveness | passed |

jasper-d commented:
Hmm, after making the requested changes I'm observing a lot of errors. One problem is that when using a LB w/ external-dns it takes time until the LB is fully setup (i.e. has a public IP) and DNS records are created. More worringly, MongoDB instances report the following error:

`E  REPL     [conn418] replSetReconfig got NewReplicaSetConfigurationIncompatible: New and old configurations both have members with host of mycluster-rs-0.db.example.com:27017 but in the new configuration the _id field is MemberId(1) and in the old configuration it is MemberId(0) for replica set rs while validating { _id: "rs", version: 2, members: [{ _id: 1, host: "mycluster-rs-0.db.example.com", arbiterOnly: false, buildIndexes: true, hidden: false, priority: 1, tags: { serviceName: "mycluster" }, slaveDelay: 0, votes: 1 }, { _id: 2, host: "mycluster-rs-1.db.example.com", arbiterOnly: false, buildIndexes: true, hidden: false, priority: 1, tags: { serviceName: "mycluster" }, slaveDelay: 0, votes: 1 }, { _id: 3, host: "mycluster-rs-2.db.example.com", arbiterOnly: false, buildIndexes: true, hidden: false, priority: 1, tags: { serviceName: "mycluster" }, slaveDelay: 0, votes: 1 } ], protocolVersion: 1, settings: { chainingAllowed: true, heartbeatIntervalMillis: 2000, heartbeatTimeoutSecs: 10, electionTimeoutMillis: 10000, catchUpTimeoutMillis: -1, getLastErrorDefaults: { w: 1, wtimeout: 0 }, replicaSetId: ObjectId('5ede3587f19adbcb3eb3d86e') }, writeConcernMajorityJournalDefault: true }  `

I don't have the time to investigate this any further, but would recommend to not merge the PR.

delgod commented:
@jasper-d because of the described issue, we have `sleep` command in `getExtServices` function 😉
`getExtServices` waits till LB creates DNS records and IPs
